python/triton/language/semantic.py

```diff
@@ -1543,3 +1563,71 @@ def advance(base: tl.tensor, offsets, builder: ir.builder) -> tl.tensor:
 
     # Advanced block pointer type is the same as before
     return tl.tensor(builder.create_advance(base.handle, offsets), base.type)
+
+# ===----------------------------------------------------------------------===##
+# Utilities for symbolic shape propagation
+# ===----------------------------------------------------------------------===##
+
+
+_DimT = Union[int, "tl.constexpr_placeholder", "tl.tensor"]
+
+
+def _shape_get_numel(shape: List[_DimT]) -> int:
```

Comment:
Below are just some utility functions to help deal with symbolic shapes(shape sizes mixing int and tl.tensor/tl.constexpr_placeholder)